bundles/org.openhab.binding.modbus.solaxx3mic/README.md

```diff
@@ -0,0 +1,296 @@
+
+Moreover, this binding should be really easy to extend to other devices with single block of registers, because it is required only to edit thing description xml files from resources\OH-INF folder to transform it to completelly different binding. 
+Java code is written that way, that is scans all defined and mapped to items channels and maps received modbus registers block to them dynamically.
+
+## Discovery
+
+Because of flexibility of this binding discovery is not supported.
+
+## Binding Configuration
+
+Binding does not requires any special configuration.
+
+## Thing Configuration
+
+Thing configuration is necessary in configuration file because of extended amount of information required by thing.
+
+Basically this thing requires Bridge of type Modbus TCP or Modbus Serial.
+It also requires definition of following parameters:
+
+1. refresh - amount of time between polls [s]
+2. inputAddress - begin of modbus type 3 registers block
+3. inputBlockLength - length of modbus type 3 registers block
+4. maxTries - maximal amount of tries during failure condition
+5. holdingAddress - begin of modbus type 4 registers block
+6. holdingBlockLength - length of modbus type 4 registers block
+
+
+## Channels
+
+Channels are splitted into several groups.
+All channels are read only.
+Not all channels are required to use.
+
+Groups:
+
+1. Device information
+2. AC Line General
+3. AC Phases A, B, C as separate Groups
+4. DC Lines 1 and 2 as separate Groups
+
+Channels in every group type:
+
+1. Device Information (device-information)
+   - status (status-type)
+   - cabinet-temperature (cabinet-temperature-type)
+   - heatsink-temperature(heatsink-temperature-type)
+2. AC Summary (ac-general)
+   - ac-power (ac-power-type)
+   - ac-dayly-energy (ac-dayly-energy-type)
+   - ac-lifetime-energy (ac-lifetime-energy-type)
+3. AC Phase (ac-phase)
+   - ac-phase-voltage-to-n (ac-phase-voltage-to-n-type)
+   - ac-phase-current (ac-phase-current-type)
+   - ac-phase-frequency (ac-phase-frequency-type)
+   - ac-phase-power (ac-power-type)
+4. DC summary (dc-general )
+   - dc-current (dc-current-type)
+   - dc-voltage (dc-voltage-type)
+   - dc-power (dc-power-type)
+
+Every channel configuration consists of five parameters:
+
+1. registerFunction - identification of block where register resides.
+   - Possible are:
+     - 3 - Input register
+     - 4 - Holding register
+2. registerNumber - number of register inside of block
+3. registerType - type of register:
+   - INT
+   - INT_BIGENDIAN
+   - SHORT
+   - USHORT
+   - STATUS
+4. registerUnit - unit of value in register. Possible are all names from org.openhab.core.library.unit.Units constants. It is required to use exact names with capital Letters only. It is not verified at this moment, but when coversion fails, there is an exception thrown and message to logs is written: "Illegal access exception during reflection to Units!"
```

Comment:
I haven't found any way to convert SI based unit to the org.openhab.core.library.unit.Units. So I have choosen to use reflection and use this parameter as a name of field to get by reflection.
That is the reason of java details here.
I assume it is most elegant way to be consistent.

Reply:
You should not use this method. The Units class is an implementation detail and no official api. And as you mention it doesn't work with SIUnits. The way to do this is to pass the value + unit as a string to QuantityType. So if the user specifies the unit just concatenate the value + unit. For example `"123kW"`

Reply:
@Hilbrand - I am using QuantityType just during the scaling procedure.
But first I have to gather all information required.
Those are:
1. Raw Value it self
2. Source Unit
3. DestinationUnit
Source and destination units must differs only in preffix.

So I have created following:
```
    protected State getScaled(Number value, int scaleFactor, Unit<?> unit) {
        if (scaleFactor == 0) {
            return new QuantityType<>(value.longValue(), unit);
        }
        return new QuantityType<>(BigDecimal.valueOf(value.longValue(), scaleFactor * -1), unit);
    }
//////////////////////////////////////////////////////////////////////////////////////////////////
    Field field = Units.class.getDeclaredField(solaxChannelConfig.registerUnit);
    Unit<?> unit = (Unit<?>) field.get(field.getClass());
    if (unit != null) {
        State s = getScaled(value, solaxChannelConfig.registerScaleFactor, unit);
        // logger.debug("value of channel is {} (real value = {}, scaleFactor = {}, unit = {}",
        // s.toString(), value, solaxChannelConfig.registerScaleFactor, unit.toString());
        updateState(localchannel.getUID(), s);
```

Could you point me out how to correct this?

Reply:
If you let the user specify the unit (e.g. `kW`) itself instead of the `Unit` name you would not need the scaling factor. Because the user can just specify the unit of the source data. That unit is passed to QuanitityType. So it would just be something like `new QuantityType<>(value + unit)` Where value is a `float` or `double` and unit is a `String`. When it is in a QuantityType the user can specify in the item what unit it should display. So if the source data is in watt `W`. The user can configure the item to display in `kW` and the framework will do the conversion so it actually displays in `kW` even while the input was in `W`.

Reply:
How does `QuantityType` recognizes source unit? I think I understand your point, could you review this snippet of code?
```
float rawData = 5022.0 ; // it means inverter produced 502.2Wh this day, we want to have it kW
string sourceUnit = "dW";
string finalUnit = "kW" ;
State s = new QuantityType<>(rawData + sourceUnit, finalUnit) ;
```

Reply:
You will construct quantity type using the source number and source unit. There is no need to handle destination unit in code.

In your example source number 5022, and source unit of "W/10"

Openhab handles the rest. User can configure display unit to be kW for example.

Reply:
Alternatively you could have parameter called "gain", e.g. "0.1 W" to convert raw numbers to watts.

output=new QuantityType(gain)*input_number